htlcswitch/payment_result.go

```diff
@@ -258,3 +258,42 @@ func fetchResult(tx kvdb.RTx, pid uint64) (*networkResult, error) {
 
 	return deserializeNetworkResult(r)
 }
+
+// cleanStore removes all entries from the store, except the payment IDs given.
+// NOTE: Since every result not listed in the keep map will be deleted, care
+// should be taken to ensure no new payment attempts are being made
+// concurrently while this process is ongoing, as its result might end up being
+// deleted.
+func (store *networkResultStore) cleanStore(keep map[uint64]struct{}) error {
+	return kvdb.Update(store.db.Backend, func(tx kvdb.RwTx) error {
+		networkResults, err := tx.CreateTopLevelBucket(
+			networkResultStoreBucketKey,
+		)
+		if err != nil {
+			return err
+		}
+
+		// Iterate through the bucket, deleting all items not in the
+		// keep map.
+		var numDeleted int
+		cursor := networkResults.ReadWriteCursor()
+		for k, _ := cursor.First(); k != nil; k, _ = cursor.Next() {
+			pid := binary.BigEndian.Uint64(k)
+			if _, ok := keep[pid]; ok {
+				continue
+			}
+
+			numDeleted++
+			if err := cursor.Delete(); err != nil {
```

Comment:
didn't we have an issue in the past where deleting while iterating caused issues? maybe i'm misremembering the issue tho...

Reply:
You're right, this is still a bug: https://github.com/etcd-io/bbolt/issues/146

When you mention it I remember we had to work around it in `btcwallet`

Reply:
Reverted back to using the original `ForEach`: https://github.com/lightningnetwork/lnd/pull/3131/commits/1ff47c65e1c789f7468a1e5dc172d4cda5dfc270

cc @bhandras